Convert statistics to use history api for database access

bdraco opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Convert statistics to use history api for database access

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Convert statistics to use history api for database access

ThomDietrich commented:
Hey @bdraco,
I suppose this is in reaction to https://github.com/home-assistant/core/pull/68315#issuecomment-1072308883 ?
While your change probably makes sense, I actually wanted to start the conversation whether the statistics component should hold its own copy of historic states in `self.states`, rather than accessing `history` directly!?

bdraco commented:
Hi @ThomDietrich 👋 

I adjusted all the integrations that were mentioned here https://github.com/home-assistant/core/pull/68315#issuecomment-1072116361 to use the history api (and adjusted the ones that were already using it that could be made more efficient)

I think the current design is fine that this integration runs `_initialize_from_database` when the entity is added to Home Assistant since it should only happen once.  The state machine will not have historic data at startup so there isn't many other  options.

bdraco committed:
Merge branch 'dev' into no_attr_if_not_needed_statistics

bdraco committed:
fix typing

emontnemery commented:
Another option would be to do something like this: <https://github.com/home-assistant/core/pull/66851>; i.e. a custom stored state which can store a buffer of states.

emontnemery commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -474,37 +472,29 @@ def _scheduled_update(now: datetime) -> None:
+            limit=self._samples_max_buffer_size,
+            include_start_time_state=False,
+        )
+        # Need to cast since minimal responses is not passed in
```
What does this comment mean?
Also, why is the return value of `state_changes_during_period` as flexible as `MutableMapping[str, Iterable[LazyState | State | dict[str, Any]]]` yet we can just cast it to `list[State]`?

bdraco replied:
If minimal_responses is True it will return the intermediate states as dicts.  In the future we could split it into two functions to clean up the typing

emontnemery replied:
What is `minimal_responses`, there's no such parameter in `history.state_changes_during_period`?

bdraco replied:
We could cast it in `state_changes_during_period` as its inheriting the typing from:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/recorder/history.py#L507

bdraco replied:
Let me know if you prefer that or keeping it as-is and I'll make the change

emontnemery replied:
Yes, let's correct the typing of `state_changes_during_period` (and  `get_last_state_changes`) since we know they will never include list elements.

bdraco replied:
https://github.com/home-assistant/core/pull/68917/files

bdraco replied:
https://github.com/home-assistant/core/pull/68917/files

bdraco replied:
Looks like it still has to be cast for now since we call reverse on it.  Once we do the refactor in the other PR to split it we can take it out

emontnemery commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -474,37 +472,29 @@ def _scheduled_update(now: datetime) -> None:
+            limit=self._samples_max_buffer_size,
+            include_start_time_state=False,
+        )
+        # Need to cast since minimal responses is not passed in
```
```suggestion
        # States are returned as State objects, safe to cast
```

bdraco committed:
Merge branch 'dev' into no_attr_if_not_needed_statistics

bdraco commented:
Marking this for 2022.4 since it was reported to cause a startup error since it doesn't go through the compat logic as is
